lib/Heap_List.thy

```diff
@@ -383,6 +383,14 @@ lemma not_head_prev_not_None:
   using sym_heap_prev_None_is_start heap_path_head
   by fastforce
 
+lemma heap_ls_neighbour_in_set:
+  "\<lbrakk>heap_ls hp st xs; sym_heap hp hp'; st \<noteq> None \<longrightarrow> hp' (the st) = None; p \<in> set xs\<rbrakk>
+   \<Longrightarrow> \<forall>nbr. (hp p = Some nbr \<longrightarrow> nbr \<in> set xs) \<and> (hp' p = Some nbr \<longrightarrow> nbr \<in> set xs)"
```

Comment:
Is the object logic `∀` necessary or could `nbr` just be a free variable?

Reply:
This could probably be changed to have a free variable. This is really only used to prove a variant that has `list_queue_relation` in the assumptions, instead of `heap_ls` (and because it has `list_queue_relation`, which currently lives in `TcbAcc_R`, I've left it in Refine, though we could maybe move that to `Lib` too if we really wanted to). That variant is used in a fair few places with `cut_tac` and `frule`, the idea being that we then have a fairly general statement sitting around in our assumptions that `fastforce` can go to town with. I think we don't want a fixed free variable for that rule, because we want to reason about both previous pointers and next pointers

Reply:
If there's a good reason, it's fine and if  it's used mostly as `drule`/`frule` and we want to delay instantiation then that sounds good to me.